### ISSUE TITLE
FYI: CanvasBlocker can be detected by sites

### ISSUE THREAD

**Issue by earthlng** [ghacksuserjs/ghacks-user.js#174]
*moved from https://github.com/ghacksuserjs/ghacks-user.js/issues/12*

https://github.com/kkapsner/CanvasBlocker/issues/114#issuecomment-295259795

Detection Test: https://kkapsner.github.io/CanvasBlocker/test/detectionTest.html

edit: other interesting links about Canvas fingerprinting:
* https://browserleaks.com/canvas#how-does-it-work
* [Websites using HTML5 canvas fingerprinting](https://web.archive.org/web/20141228070123/http://webcookies.org:80/canvas-fingerprinting/) - with code samples
* [List of websites that track you using Canvas](https://multiloginapp.com/sites-track-canvas/)
* [Canvas Fingerprinting Scripts](https://securehomes.esat.kuleuven.be/%7Egacar/persistent/#canvas-results) - thanks @Atavic 
 => shows the summary of canvas fingerprinting scripts found on the homepages of top 100K Alexa sites.
  * LOL some of them didn't even change the "Text drawn into the canvas" :man_facepalming:

**Comment by earthlng** [comment]
@zymase wrote:

> > FYI: CanvasBlocker can be detected by sites
> 
> @earthlng does your detection test concern only the _CanvasBlocker_ add-on or does it detect the fact canvas is blocked, whatever the blocker? I am wondering if the other similar FF add-on, _Canvas Defender_ would be detected by your test, and if is it detected by other sites as _CanvasBlocker_ is.

@Gitoffthelawn wrote:

> @zymase I just tested it for you. Canvas Defender is not detected.

**Comment by earthlng** [comment]
To clarify, this is not *my* detection test, I only copied it from kkapsner's repo to make it available online

**Comment by earthlng** [comment]
I don't know that other addon. If it has a "fake readout" functionality it's possibly detectable too. If it only blocks canvas readout then that's detectable anyway. This test is specifically for CanvasBlocker.
But if you look at the 3 year old bugzilla ticket that kkapsner him/her-self submitted, it mentions that any addon that uses exportFunction can be detected that way.

**Comment by zymase** [comment]
I'm carrying on this thread since @earthlng does as well ...

I'll echo my comment on Ghacks :+1: 

> I've just disabled CanvasBlocker (CB) and installed 'Canvas Defender' (CD)
> At this point I'm wondering if CD is reliable. Why?
> 
> CB with Block Mode=fake readout API has the option to notify the user when the fake mode is used.
> CD as well displays a notification.
> 
> When testing, both add-ons notified the canvas on the BrowserLeaks Canvas test page, but when testing on Google Maps, only CB notified, not CD. Google Maps definitely uses Canvas. So why did CD not notify me?
> 
> CD, 'Canvas Defender' ver. 1.1.0 is at this time marked as experimental by its developers. Maybe is that the explanation?

At this point I remain uncertain on what is the best option for managing Canvas tracking.

**Comment by Atavic** [comment]
[CB](https://github.com/kkapsner/CanvasBlocker) blocks the Javascript <canvas> API.
[CD](https://multiloginapp.com/how-canvas-fingerprint-blockers-make-you-easily-trackable/) creates a unique and persistent noise that hides your real canvas fingerprint: it's blocking vs spoofing.

> what is the best option

CB alone isn't a good option, you should find some combo as [this one](https://github.com/privacytoolsIO/privacytools.io/issues/121#issuecomment-269809320).
Are you going to block such fingerprinting APIs? Block most of them. 

Otherwise, if you go with the spoofing option, you should check how much reliable is the spoofing method. Altough spoofing sound better to me, there are [some issues](https://github.com/gorhill/uBlock/issues/1696#issuecomment-224593108) to consider.

**Comment by zymase** [comment]
As far as I understand CB, when choosing CB with Block Mode=fake readout API the process is not that of blocking, literately. CB has several modes (see https://addons.mozilla.org/en-US/firefox/addon/canvasblocker/), and,
> fake readout API: Canvas Blocker's default setting, and my favorite! All websites not on the white list or black list can use the <canvas> API to display something on the page, but the readout API is forced to return a new random value each time it is called

Moreover, as I mentioned above, CD didn't notify the active Canvas on Google Maps ...

**Comment by Atavic** [comment]
If it didn't notify, the site has bypassed CD, as Gorhill said on my link above, these specific addons face many workarounds by site owners and developers.

**Comment by zymase** [comment]
Well, the site mentioned above doesn't bypass CB ... and CB did not block the canvas (with Block Mode=fake readout API ) but spoofed it in fact, as far as I understand it. I even believe Google Maps just wouldn't tun correctly with Canvas blocked.

**Comment by Thorin-Oakenpants** [comment]
> CD didn't notify the active Canvas on Google Maps ...

May be a limitation of WE APIs

**Comment by Atavic** [comment]
Gmaps needs the [Canvas API](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial) to be working correctly. Maybe CB works in a clever way, by spoofing a readout API response that doesn't break Gmaps? better ask @kkapsner

**Comment by zymase** [comment]
I've tested Google Maps (in fact the true Canvas requirement is Google Street View) with CB's BlockMode=Block Readout API and Google Street View failed (black screen).

Now, I read at https://addons.mozilla.org/en-US/firefox/addon/canvasblocker/reviews/890823/ the developer of CB answering this:

> You could switch to "fake input"-mode which should not be detected. But this mode is less secure - fingerprinting over WebGL is possible with that. 

So if Goofle Maps/Street View runs with CB/Block Mode=fake readout API it's maybe because, as Ataviv mentioned it, because that mode is bypassed by Google.

In fact when a Canvas blocker/defender doesn't break a Canvas page it is either because it's efficient either because the page bypassed it. This is cruel.

**Comment by zymase** [comment]
@Atavic wrote,

> Maybe CB works in a clever way, by spoofing a readout API response that doesn't break Gmaps?

That's what I had in mind, a hope rather than a certitude, because of above mentioned.

All this brain storming for the sole purpose of trying to defeat Web tracking....

**Comment by zymase** [comment]
With CanvasBlocker / Block Mode=fake readout API , testing on;

1- https://browserleaks.com/canvas
![screengrab-2017-07-17-231536](https://user-images.githubusercontent.com/16312589/28290433-b36520ae-6b46-11e7-94b0-dae2e2c10b4b.png)

-> Canvas signature different with every page reload

2-https://browserleaks.com/proxy
![screengrab-2017-07-17-231612](https://user-images.githubusercontent.com/16312589/28290478-e8c3ecf8-6b46-11e7-8426-cdb096f14646.png)

Yet, CanvasBlocker IS detected by the above CanvasBlocker tester page ...

**Comment by Atavic** [comment]
This particular fingerprinting isn't so widely used and most of the sites using it call a Javacript file hosted by `addthis.com`.

See: https://securehomes.esat.kuleuven.be/~gacar/persistent/#canvas-results

uBlock's Privacy filter list blocks it.

**Comment by Thorin-Oakenpants** [comment]
@zymase  https://browserleaks.com/proxy (your second image)

With CB & fake readout API I get
> HTML5 Canvas Protection	✔ not detected 

This is at complete odds with your test

**Comment by Thorin-Oakenpants** [comment]
@Atavic said
> CB on default uses spoofing as default, but I use it on block mode.

That's a really great way to raise your entropy. Canvas fingerprints are *unique* - they don't narrow you down, they actually identify you. You have two options - spoof or block. Spoofing means you are always unique, but never tracked by this metric each time it is gathered. Blocking by contrast means you give a constant value. Can you see where I am going with this?

Add the fact that SFA (sweet f**k all) people block, so you're now in a very small subset. Why do I say SFA - because 90% or more of users will not bother to do anything about it, and of the rest, most will use default, or most will know that blocking is shit. I do not have metrics - I could try to clock on panopticlick (but then the web page script stops working), or some other FP test site, but these sites are heavily skewed. I guestimate you're under 1%.

Spoofing is the best option but needs work. If your connected IP is generating a random canvas value for each request, you also stick out like a dog's balls (if they want to track you across a domain). A better option is a random value per domain per session - done automagically.

Used up way more than my 2cents

**Comment by zymase** [comment]
@Thorin-Oakenpants ,

> With CB & fake readout API I get
>     HTML5 Canvas Protection ✔ not detected
> This is at complete odds with your test

I did mention,

> In fact, HTML5 Canvas Protection is detected when reloading that test page but not on first load.

When the test is first performed HTML5 Canvas Protection **is not** detected,
When I reload the test page, HTML5 Canvas Protection **is** detected.

**Comment by crssi** [comment]
^^I have `not detected` on reload also.
Using CB, fake readout API, persistent

**Comment by crssi** [comment]
But the link provided by @earthlng gives me `function length: CB detected`

**Comment by zymase** [comment]
@crssi ,

> ^^I have not detected on reload also.
> Using CB, fake readout API, persistent

Tested again, Block mode = fake readout API
1- Random number generator = non persistent -> Not detected on initial page, detected on page reload
2- Random Number generator = persistent -> detected on initial page load (+ of course on reload).

This is mystery. Of course, as you, earthlng's link to CanvasBlocker detection test proves CB detected.

Reminds me my teen-age years, wondering if she loved me or not :)

**Comment by kkapsner** [comment]
To bring a little bit insight to the question if the canvas protection is detected or not: this protection detection is performed on a one pixel canvas. CB adds exactly one bit of entropy to one pixel. Therefor it's 50% that the pixel is not altered and therefore the detection does not work.

I will update my detection test to include this tests.

**Comment by crssi** [comment]
@kkapsner to the rescue... nice to have you here. :)

**Comment by earthlng** [comment]
@kkapsner hi and thanks for chiming in here :)

> this protection detection is performed on a one pixel canvas.

"**this** protection detection" - which one? https://browserleaks.com/proxy ?

> I will update my detection test to include this tests.

I hope you don't mind that I copied your test file and hosted it on my repo to make it accessible online.
I tried a couple of links to see if you had already set it up on your repo but it didn't seem like it.
Maybe you could change the settings on your repo to make your test folder online accessible via a github.io page, that would be great.

**Comment by kkapsner** [comment]
Yes - the one from browserleaks.

I don't mind you copying it. But you're right having them online available ist better: http://kkapsner.github.io/CanvasBlocker/test/detectionTest.html

**Comment by Atavic** [comment]
@zymase Seems like the test page compares the results:

> When the test is first performed HTML5 Canvas Protection is not detected,

Test page stores your fingerprint signature

> When I reload the test page, HTML5 Canvas Protection is detected.

Test page compares the results and the only value that has changed raises the detection.

**Comment by zymase** [comment]
@Atavic that's what I had in mind initially. I just checked, as before with **Block Mode = fake readout API** _BUT_ with this time **Random Number generator = persistent** => same sig and indeed Canvas Protection _not_ detected, initially and on page reload.

Makes me realize I mistaked above when I answered to crssi:
> 1- Random number generator = non persistent -> Not detected on initial page, detected on page reload
> 2- Random Number generator = persistent -> detected on initial page load (+ of course on reload).

In fact it is:
2- Random Number generator = persistent -> NOT detected on initial NOR on reload.

What likely happened is that I had forgotten to clear the site's cache/cookie.

**Comment by earthlng** [comment]
@kkapsner Great, thanks, I updated the link in my first comment and will remove the now outdated detection test from my repo again.
I see that you added more tests in the meantime. Hopefully you'll not try to fix the "known pixel value" tests, because you can never known which part of a canvas a fingerprinting site would look at, right?
I also thought of another one that illustrates another problem, one that Canvas Defender tries to prevent. double readout:
```js
var readout1=ctx.getImageData(0, 0, 10, 10).data;
var readout2=ctx.getImageData(0, 0, 10, 10).data;
if (readout1 !== readout2) { console.log("blocker detected", readout1, readout2); }
```
I'm not sure comparing arrays that way is the correct way to do it but it seems to work and detects CB's fake readout mode

**Comment by earthlng** [comment]
> Seems like the test page compares the results

@kkapsner already explained that - it's pure RNG ie sometimes CB happens to change the value the code is looking at and other times not. 50-50 chance

This is the code the page uses, which shows that they only look at the "red" value (RGBA) and if that changes it knows that the canvas was tampered with
```js
function tbb_canvas() {
  try {
    var a = document.createElement('canvas').getContext('2d');
    return a.fillStyle = 'rgb(3,3,7)',
    a.fillRect(0, 0, 1, 1),
    a.fill(),
    3 !== a.getImageData(0, 0, 1, 1).data[0]
  } catch (a) {
    return !1
  }
}
```

**Comment by earthlng** [comment]
@kkapsner how much identifiable bits can a single pixel really give? Not much I reckon. You could maybe detect if a site only wants ImageData for width=1 && height=1 and not tamper with that. That would also always return correct values for any [color pickers](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Pixel_manipulation_with_canvas#A_color_picker) out there. Like? xD

**Comment by earthlng** [comment]
>  I assume most sites that use canvas fingerprinting have to use a larger canvas than that

https://web.archive.org/web/20141228070123/http://webcookies.org:80/canvas-fingerprinting/
Seems like they all copy from each other, look at all the `,62,20` in there.

https://browserleaks.com/canvas#how-does-it-work
browserleaks also happens to use 62,20 for the canvas size. Not that you should limit it to that, at all, but excluding 1,1 seems only beneficial.

**Comment by earthlng** [comment]
Nice, but I just realized that nothing really stops a site from loading a larger canvas pixel by pixel and putting the array together that way, right? Damn! Maybe make the minimal canvas size optional?

> The double readout detection can be circumvented by using the persistent random number generator.

oh nice, I was wondering if that's what that is for. How does that work - does it keep the same data across all windows and tabs for the whole session, and unless "stored" will only change with the next FF startup?
I'm going to use that then, but why would someone want to store it though?

**Comment by kkapsner** [comment]
I have to correct me: every pixel has four bits of entropy - one for every channel.

**Comment by kkapsner** [comment]
I changed the way the max readout size works: in the new version (unfortunatelly I have no release date or schedule as I'm waiting for Mozilla to provide an API so I can feed the settings to the tabs before the scripts of the web page run - hoorray WebExtensions!!!) it's only dependent on the canvas size and not on the readout size. Same for the min size - so a large canvas can not be put together by single pixels.
But as most settings you can turn it off  - just set it to zero.

Yes - the persistent rng keeps the same data across all windows and tabs for the whole session. If you tick the "store persistent data" it will be written to your settings and reloaded after shutdown.
But keep in mind that you can be tracked then within one session - but only within one domain.

**Comment by earthlng** [comment]
hoorray WebExtensions!!! indeed - good stuff 🤦‍♂️ 

> so a large canvas can not be put together by single pixels.

Why not? I don't see how you could possibly prevent that, assuming you don't simply ignore how much data a site requested.

> Yes - the persistent rng keeps the same data across all windows and tabs for the whole session.

Thanks for confirming. I also noticed on the [browserleaks testpage](https://browserleaks.com/canvas) that if you look at the image details, the file size and number of colors barely changes with "persistent" but with "non persistent" it changes a lot. So that probably stands out a lot. The file size with CB disabled or CB "persistent" is around 2kb and with non persistent it's upwards of 7kb, while the # of colors changes from around 70 to 300+. That and the fact that if your IP doesn't change neither should the canvas fingerprint (IMO), I think making "persistent" the default setting would probably be better. I think even in Tor Browser "persistent" would be preferable. Thanks for creating and maintaining this great addon btw!

**Comment by earthlng** [comment]
@Gitoffthelawn wrote:

> @zymase I just tested it for you. Canvas Defender is not detected.

That's weird because when I tested it, it was detected by both length and name. Not only that but its exposed function also had 2 additional variables that neither the original function nor CB's function has.
Did you create a copy of the file and test it via file:/// ?
Canvas Defender tries to inject some code into a page and that can be blocked by a site with CSP. Idk exactly what it tries to inject but I do know that github.io doesn't allow script:self and therefore blocks the injection. Maybe that's why we got different results.
They also advertise one of their products in CD so I definitely know which one of the 2 addons I prefer.

**Comment by zymase** [comment]
> Canvas Defender is not detected.

@earthlng are we referring to the same test?
The Canvas detection test I and @Gitoffthelawn were referring to was
https://kkapsner.github.io/CanvasBlocker/test/detectionTest.html
On that test page _CanvasBlocker_ was detected, _Canvas Defender_ was not.

But running _Canvas Defender_ and testing on https://browserleaks.com/proxy showed that a Canvas Blocker was indeed detected.

**Comment by earthlng** [comment]
> are we referring to the same test?

yes, I tested it yesterday and CD was detected. Idk what to tell you, no idea whats causing that.
Do you get a notification from CD on that page?

**Comment by zymase** [comment]
> Do you get a notification from CD on that page?
I assume you mean on https://kkapsner.github.io/CanvasBlocker/test/detectionTest.html

I disabled CB and installed CD (again) this morning to test CD on https://browserleaks.com/proxy and Canvas Blocker was detected. I did not test this time on https://kkapsner.github.io/CanvasBlocker/test/detectionTest.html because previous testing there showed CD was not detected.

If you wish I may test again CD on https://kkapsner.github.io/CanvasBlocker/test/detectionTest.html
Meanwhile I have removed CD to stick on CB.

**Comment by Just-me-ghacks** [comment]
https://kkapsner.github.io/CanvasBlocker/test/detectionTest.html
CanvasBlocker 0.3.8 - detected
Canvas Defender 1.1.0 - detected

**Comment by zymase** [comment]
I've removed the _BlockCanvas_ Firefox add-on and wont be using _Block Defender_ either.

Why? Seems to me this blocking is pertinent when the scenario is using multiple identities and/or several browsers. If I understand correctly what is explicitly explained by the developer of [Block Defender](https://multiloginapp.com/how-canvas-fingerprint-blockers-make-you-easily-trackable/) : 

> **The only viable solution that exists**
> 
> Make the canvas fingerprinting function available on the websites you visit. (So it’s not clear you are wearing a mask.)
> Use a canvas identity with consistency. (So it’s not clear you are trying to avoid detection.)
> Switch up the identity when necessary. (To erase your tracks.)

I get to consider that blocking Canvas fingerprinting has no real advantage when other defenses aren't included, in other words don't block the fingerprinting unless you change identity. I'm not concerned.

**Comment by earthlng** [comment]
@zymase "multiple identities" can also mean different IPs. If your IP changes but you still produce the same (perhaps unique) canvas fingerprint, that's something you may not want.

**Comment by zymase** [comment]
@earthlng , 

> "multiple identities" can also mean different IPs. If your IP changes but you still produce the same (perhaps unique) canvas fingerprint, that's something you may not want.

Yeah, that's how I get it. My feeling is that Canvas Blocking fills the gap left when the scheme is that of multiples identities or TOR, VPN : the user hides but is recognized by a Canvas fingerprinting : too stupid, all that work to finally get spotted with a Canvas! Otherwise, a simple traveler like me ends up being spotted as someone who breaks Canvas fingerprinting without the benefit of not being recognized ...

Again, nothing is definitive. I try to learn and if I understand that I'd be mistaking that I'd reconsider Canvas fingerprinting when applied for my Web sessions.

**Comment by Thorin-Oakenpants** [comment]
I added the test detection to the [Wiki](https://github.com/ghacksuserjs/ghacks-user.js/wiki/Appendix-C:-Test-Sites)

**Comment by Thorin-Oakenpants** [comment]
Even if you use canvas consistently when spoofing, you'll still get picked up probably if they really wanted to track you. Not that we don't want to make it hard for them. I actually think there is so much lower hanging fruit and loads of plebs with no clue, that tweakers like us are not the target. Tracking _changes_ in canvas is likely to be very very very low - slim to none.

Hopefully at some stage in the future, the entire firefox base :middle_finger: 's canvas with an enforced spoof. Seriously, what good is it for - its not a webcompat thing. Just make the default to always spoof for the current 100million FF users
